### DISCUSSION TITLE
Terraform bootstrap

### DISCUSSION THREAD

**Discussion by kaihendry** [argoproj/argo-cd#9865] [Q&A]
Assuming I have a typical AWS VPC, EKS Terraform, how do simply bootstrap ArgoCD?

Is there some official Terraform module?

Idea is if I ever needed to recreate the EKS, ArgoCD is quickly bootstrapped and brings up the cluster...

**Comment by mtrin** [comment]
I'd say the [operator](https://argocd-operator.readthedocs.io/en/latest/) would be the recommended approach.
From there I believe the best approach is [Kustomize](https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/#manage-argo-cd-using-argo-cd)
Terraform would be more for managing your EKS, AWS etc.

**Reply by kaihendry** [reply]
Is there some Github workflow to provision this Operator? That's the missing link for me!

**Reply by mtrin** [reply]
It's not meant to have, this is what operators are trying to solve. You need something to manage that CRD and so on. It's a chicken and egg thing.
This is how I understand anyway, I should add that I'm not part of the ArgoCD team but am implementing at work.
Edit: Kustomize as I understand manages Argo CD after bootstrapping.

**Comment by DpoBoceka** [comment]
I'm currently having a similar issue with insignificant differences, only I'm using [pulumi](https://www.pulumi.com/) instead of terraform and the target infrastructure is in GCP.  
My approach is:
1. Deploy a "infrastracture-control-plane" k8s cluster and attach nodepool to it via IaC
3. Create a k8s secret with GH PAT in the cluster via IaC. I store the secret itself in Google Cloud Secret Manager, so terraform/pulumi get the payload of a secret from there.
4. Create namespace for ArgoCD
5. Use [kustomize resourse](https://www.pulumi.com/registry/packages/kubernetes/api-docs/kustomize/) of pulumi's k8s provider

And that's all I want from pulumi/terraform.
Then ArgoCD deploys [crossplane](https://crossplane.io/) to create other resources in GCP(including other k8s clusters), `external-secrets`, `grafana-agent`, etc and takes care of itself, too.

The little trick here is to configure terraform/pulumi in a way to ignore `kustomized-argocd` resource if it's created because I want argocd to be responsible for manifests, not IaC tool

**Reply by maulerjan** [reply]
Hi, I'm dealing with the last step:

> The little trick here is to configure terraform/pulumi in a way to ignore kustomized-argocd resource if it's created because I want argocd to be responsible for manifests, not IaC tool

How did you manage to make it work? I can't seem to make Pulumi ignore my resource changes to ArgoCD itself (they are picked up by ArgoCD but Pulumi then registers that as a change and wants to re-deploy the whole thing).

**Comment by mtrin** [comment]
You guys might want to have a look on argocd autopilot as well

**Reply by EdDanileyko** [reply]
Are you aware if this project is easily deployable onto EKS via terraform?

**Reply by endophage** [reply]
@EdDanileyko I'm using autopilot on EKS but terraform has quite a few issues with Kubernetes in general. First up, you're meant to use separate `terraform apply` commands to create the EKS cluster, and deploy initial services to it. Secondly, and this depends on exactly _how_ you try and deploy via terraform, the kubernetes provider can't create and use a CRD in the same deployment.

Rather than spending days maybe making terraform work, I went with using autopilot outside of terraform. In the case of disaster recovery, autopilot can install from your gitops repo to restore an existing argocd deployment.

One quirk I did notice is that even in a recovery situation, autopilot deploys the vanilla, un-customized base argocd install, then sets up your repo as the gitops repo, triggering an immediate sync on the argocd instance to bring up your specific configuration.

**Reply by EdDanileyko** [reply]
Thanks for your reply :)
as a preface.. I am new to all of this stuff so hopefully what I am about to say makes some degree of sense.

I have separated deployment of networking resources, eks, and services running in eks across a few repos so the first challenge is less of a problem (not sure if best practice but it is how it is).

I was not aware of your point about the Kubernetes provider. Do you know why that is?

I sort of managed to get this working, in that I was able to deploy the ArgoCD-autopilot manifests using the kubectl provider after initially deploying argo via the helm provider, but somehow i am getting my pods into CrashLoopBackOff state because _maybe_ it is somehow creating two Argo deployments competing for ports?

**Reply by endophage** [reply]
Unfortunately I don't have the specifics on why the kubernetes provider can't create and use a CRD in the same `apply`. I ran into it recently trying to add fluentd (I was trying to use a CRD terraform was also creating to load some secrets for fluentd) to my terraform and ultimately just had ArgoCD deploy fluentd instead of terraform.

You should be able to see if there are 2 competing ArgoCD instances. `kubectl get deployment -A` will list all deployments across all namespaces. If each of the ArgoCD services only appears in there once then you're probably OK on that front.

If you only have the one instance, CrashLoopBackOff indicates the containers are being run and are rapidly crashing. You should be able to get some logs out which might give you more info on why it's crashing.  A tool like `kubetail` which makes it easy to grab logs from all replicas in a deployment may help there if the containers are coming up and down quickly.

**Reply by EdDanileyko** [reply]
using k9s and i see lots of this in the logs across a number of pods:
`time="2022-08-15T23:05:56Z" level=fatal msg="listen tcp :8080: bind: address already in use"`
`time="2022-08-15T23:06:16Z" level=error msg="error listening on :8080: listen tcp :8080: bind: address already in useunable to start manager" `
`1:M 15 Aug 2022 23:06:13.055 # Warning: Could not create server TCP listening socket *:6379: bind: Address in use`

it's odd, because i have 2 containers per pod, one prefixed with argocd and the other not, running two different image versions
![image](https://user-images.githubusercontent.com/22099635/184736811-6eb77ec3-6276-48f7-9bd3-35eba713d004.png)


I haven't found a heck of a lot of documentation on how to use terraform to deploy eks and argo then have it just manage itself without much manual intervention. was super excited about autopilot when i found it.

**Reply by endophage** [reply]
You definitely seem to have 2 instances of ArgoCD running there, that are even on different versions.

I've found it to be a pretty good experience to use terraform to just bring up the EKS cluster then using Autopilot to install ArgoCD and from that point doing everything in Argo. 

The main point where you see the disconnect is in creating IAM roles for service accounts or certificates for ingress. In terraform you'd be able to get the ARNs and template them into the k8s manifests. However, role ARNs shouldn't ever change once created and certificate ARNs change infrequently so managing those hasn't taken much effort.
